Initial Creation (#1)
mattBrzezinski/LambdaMaker.jl: README.md

# LambdaMaker.jl
This package is intended to generate all the boilerplate code required to deploy an AWS Lambda function via AWS SAM.

## Prerequisites

> You will to download and install the following:

Reviewer: ```suggestion
You will need to download and install the following:
```

- [AWS SAM](https://aws.amazon.com/serverless/sam/)
- [Docker](https://www.docker.com/)

## How to generate the boilerplate code
You will need to generate a new project by running,

```julia
using LambdaMaker

create_lambda_function("project_name")
```

If a directory already exists with the `project_name`, it will not be overwritten.

> You can pass `force=true` as a `kwarg` to overwrite.

Reviewer: ```suggestion
If a directory already exists with the `project_name`, an error is thrown and the directory is left untouched.
You can pass `force=true` as a `kwarg` to overwrite it.
```
to make it clear that the scripts does nothing in that case (as opposed to e.g. add files to the directory)

This will create a new directory in your working directory with the following files:
- `bootstrap`: This is the file which Lambda will use when it first runs
- `Dockerfile`: Defines the Docker image that will run in the Lambda environment
- `handle_requests.jl`: Reads the Lambda invocations, calls your code, and writes the responses

> - `Project.toml`: Basical TOML for defining your package to be deployed on Lambda

Reviewer: ```suggestion
- `Project.toml`: Basic TOML for defining your package to be deployed on Lambda
```

The entrypoint for Lambda will be the `handle_event(event_data, headers)` function, you will need to modify this to call the functions you wish to invoke.

## Workflow for building, testing, and deploying a Lambda function
AWS SAM documentation can be found [here](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/what-is-sam.html).
The short-handed version is,

Build your Docker image as:
```bash
sam build
```

Test locally by running:
```bash
sam local invoke JuliaFunction
```

If your function requires some form of input see the appendix for more information.

When you are satisfied and want to deploy your function,
```bash
sam deploy --guided
```

If you want to update your function, you can use [SAM sync](https://aws.amazon.com/blogs/compute/accelerating-serverless-development-with-aws-sam-accelerate/).

## Appendix

### Other resources
- [tk3369/aws-lambda-container-julia](https://github.com/tk3369/aws-lambda-container-julia)
### Passing input to your function
Passing inputs is usually done in the JSON format.
You will need to add a package to parse this information, such as [JSON3.jl](https://github.com/quinnj/JSON3.jl).

> After generating the boilercode, adding in your JSON parsing functionality you will need to modify the `src/project_name.jl`.

Reviewer: ```suggestion
After generating the boilerplate code and adding in your JSON parsing functionality you will need to modify the `src/project_name.jl`.
```
haven't heard boilercode before? Though I must say it sounds good 😄

> Modify the `handle_event()` function as,.

Reviewer: ```suggestion
Modify the `handle_event()` function as,
```
